src/frontend/metrics/DeviceDiagnosticMetrics.ts

```diff
@@ -0,0 +1,198 @@
+  supportedCpuArchitectures?: string[];
+  totalMemory?: number;
+};
+
+function deviceTypeToString(deviceType: null | Device.DeviceType): string {
+  switch (deviceType) {
+    case Device.DeviceType.PHONE:
+      return 'PHONE';
+    case Device.DeviceType.TABLET:
+      return 'TABLET';
+    case Device.DeviceType.DESKTOP:
+      return 'DESKTOP';
+    case Device.DeviceType.TV:
+      return 'TV';
+    case null:
+    case Device.DeviceType.UNKNOWN:
+    default:
+      return 'UNKNOWN';
+  }
+}
+
+function getMetricsDeviceId(): string {
+  const result = storage.getString('MetricsDeviceId');
+  if (result) return result;
+
+  const newId = uint8ArrayToHex(getRandomBytes(16));
+  storage.set('MetricsDeviceId', newId);
+  return newId;
+}
+
+function setIfNotNull<ObjT, KeyT extends keyof ObjT>(
+  obj: ObjT,
+  key: KeyT,
+  value: null | ObjT[KeyT],
+): void {
+  if (value !== null) obj[key] = value;
+}
+
+async function generateDeviceDiagnosticMetricsData(): Promise<DeviceDiagnosticMetricsData> {
+  const screen = Dimensions.get('screen');
+
+  const result: DeviceDiagnosticMetricsData = {
+    type: 'device diagnostics v1',
+    os: Platform.OS,
+    screen: {
+      width: screen.width,
+      height: screen.height,
+      pixelRatio: PixelRatio.get(),
+    },
+    deviceType: deviceTypeToString(Device.deviceType),
+    monthlyDeviceHash: await getMonthlyHash(
+      'device diagnostics',
+      getMetricsDeviceId(),
+      new Date(),
+    ),
+  };
+
+  setIfNotNull(result, 'brand', Device.brand);
+  setIfNotNull(result, 'manufacturer', Device.manufacturer);
+  setIfNotNull(result, 'model', Device.modelId || Device.modelName);
+  setIfNotNull(
+    result,
+    'supportedCpuArchitectures',
+    Device.supportedCpuArchitectures,
+  );
+  setIfNotNull(result, 'totalMemory', Device.totalMemory);
+
+  const isEmulator = !Device.isDevice;
+  if (isEmulator) result.isEmulator = true;
+
+  return result;
+}
+
+const isAppActive = (appState: AppStateStatus): boolean =>
+  appState === 'active';
+
+const hasEnoughTimeElapsed = (): boolean => {
+  const lastSentAtMs = storage.getNumber(STORAGE_KEY) ?? -Infinity;
+  const lastSentAtDate = new Date(lastSentAtMs);
+  return (
+    !isDateValid(lastSentAtDate) ||
+    !isSameUtcMonthAndYear(lastSentAtDate, new Date())
+  );
+};
+
+const getMillisecondsUntilNextMonth = (): number => {
+  const now = new Date();
+  return beginningOfNextMonthUtc(now).valueOf() - now.valueOf();
+};
+
+export class DeviceDiagnosticMetrics {
+  #isEnabled = false;
+  #isOnline = false;
+  #isAppActive = isAppActive(AppState.currentState);
+  #isCurrentlySendingMetrics = false;
+
+  #subscriptionCleanupFns: null | (() => void)[] = null;
+
+  setEnabled(isEnabled: boolean): void {
+    this.#isEnabled = isEnabled;
+    this.#update();
+  }
+
+  async #update(): Promise<void> {
+    this.#updateSubscriptions();
+
+    const shouldSendMetrics =
+      this.#isEnabled &&
+      this.#isOnline &&
+      this.#isAppActive &&
```

Comment:
Read this from AppState.currentState rather than storing it as state - there is no additional cost to reading this, and keeping the two states in sync just makes things difficult for ourselves (e.g. currently there are edge-cases where this could not be the same as `AppState.currentState` if setEnabled() is called at certain times in the app lifecycle, or the class is initialized at certain times).

Reply:
Changed in f94bdc53aca2fee09b3144baffd35805233f051b.